src/steamship/agents/schema/message_selectors.py

```diff
@@ -29,20 +30,34 @@ def is_assistant_message(block: Block) -> bool:
     return role == RoleTag.ASSISTANT
 
 
+def is_assistant_function_message(block: Block) -> bool:
+    is_function_selection = get_tag(block.tags, kind=TagKind.FUNCTION_SELECTION)
+    return is_assistant_message(block) and is_function_selection
+
+
+def is_user_history_message(block: Block) -> bool:
+    return is_user_message(block) or (
+        is_assistant_message(block) and not is_assistant_function_message(block)
+    )
+
+
 class MessageWindowMessageSelector(MessageSelector):
     k: int
 
     def get_messages(self, messages: List[Block]) -> List[Block]:
         msgs = messages[:]
         msgs.pop()  # don't add the current prompt to the memory
-        if len(msgs) <= (self.k * 2):
-            return msgs
+        history_msgs = [
+            msg for msg in msgs if is_user_history_message(msg)
+        ]  # filter to only user history messages
+        if len(history_msgs) <= (self.k * 2):
+            return history_msgs
 
         selected_msgs = []
         limit = self.k * 2
-        scope = msgs[len(messages) - limit :]
+        scope = history_msgs[len(history_msgs) - limit :]
         for block in scope:
-            if is_user_message(block) or is_assistant_message(block):
+            if is_user_history_message(block):
```

Comment:
probably don't need this check anymore.

Reply:
I think it depends on /why/ someone is calling get_messages?

E.g. if I just want the chat transcript to prepare a chat prompt, I'd still want these kinds of things filtered out, but if I want things like status messages or function calls, then I'd want it.